state/upgrades.go

```diff
@@ -518,84 +518,109 @@ func addNonDetachableStorageMachineId(st *State) error {
 	if err != nil {
 		return errors.Trace(err)
 	}
+
+	getPool := func(d bson.M) string {
+		var pool string
+		info, _ := d["info"].(bson.M)
+		params, _ := d["params"].(bson.M)
+		if info != nil {
+			pool = info["pool"].(string)
+		} else if params != nil {
+			pool = params["pool"].(string)
+		}
+		return pool
+	}
+
+	var needsUpgradeTerm = bson.D{
+		{"machineid", bson.D{{"$exists", false}}},
+		{"hostid", bson.D{{"$exists", false}}},
+	}
 	var ops []txn.Op
-	volumes, err := sb.volumes(
-		bson.D{{"machineid", bson.D{{"$exists", false}}}},
-	)
-	if err != nil {
+
+	volumeColl, cleanup := st.db().GetCollection(volumesC)
+	defer cleanup()
+
+	var volData []bson.M
+	err = volumeColl.Find(needsUpgradeTerm).All(&volData)
+	if err != nil && err != mgo.ErrNotFound {
 		return errors.Trace(err)
 	}
-	for _, v := range volumes {
-		var pool string
-		if v.doc.Info != nil {
-			pool = v.doc.Info.Pool
-		} else if v.doc.Params != nil {
-			pool = v.doc.Params.Pool
-		}
-		detachable, err := isDetachableVolumePool(sb, pool)
+
+	volumeAttachColl, cleanup := st.db().GetCollection(volumeAttachmentsC)
+	defer cleanup()
+
+	for _, v := range volData {
+		detachable, err := isDetachableVolumePool(sb, getPool(v))
 		if err != nil {
 			return errors.Trace(err)
 		}
 		if detachable {
 			continue
 		}
-		attachments, err := sb.VolumeAttachments(v.VolumeTag())
-		if err != nil {
+
+		var attachInfo []bson.M
+		err = volumeAttachColl.Find(bson.D{{"volumeid", v["name"]}}).All(&attachInfo)
+		if err != nil && err != mgo.ErrNotFound {
 			return errors.Trace(err)
 		}
-		if len(attachments) != 1 {
+		if len(attachInfo) != 1 {
 			// There should be exactly one attachment since the
 			// filesystem is non-detachable, but be defensive
 			// and leave the document alone if our expectations
 			// are not met.
 			continue
 		}
-		machineId := attachments[0].Machine().Id()
+		machineId := attachInfo[0]["machineid"]
+		if machineId == "" {
+			machineId = attachInfo[0]["hostid"]
+		}
 		ops = append(ops, txn.Op{
 			C:      volumesC,
-			Id:     v.doc.Name,
+			Id:     v["name"],
 			Assert: txn.DocExists,
 			Update: bson.D{{"$set", bson.D{
 				{"machineid", machineId},
```

Comment:
I was expecting you to set it to hostid here. Or is that coming in an upgrade step further down?

Reply:
Right, we need to go through the steps. These 2.2 steps should behave as before to be safe.